* We will investigate the housing price dataset from Bay Area Home Sales Database and Zillow to Predict The Bay Area’s Home Prices.
* This dataset was based on the homes sold between January 2013 and December 2015.
* Please use all the techniques we have learned in the class to preprocesss/clean the dataset.
* You can drop redundant features.
<p style="color:blue"><b>final_data.csv</b></p>
* Please apply the folowing two methods to Predict The Bay Area’s Home Prices.

<h3>Method 1: Multiple Linear Regression</h3>
* Bulding the optimal model using Automatic implementations of Backward Elimination
```
def backwardElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j] == maxVar):
                    x = np.delete(x, j, 1)
    regressor_OLS.summary()
    return x
```
* <mark>Split the dataset into training sets and test sets</mark>
* Fit LinearRegression to the training sets 
* Print the Linear Regression R squared score on the test sets
* Print the Linear Regression RMSE score on the test sets

<h3>Method 2: Least Absolute Shrinkage and Selection Operator(Lasso)</h3>
* Fit Lasso to the training sets. 
* Print the Lasso R squared score on the test sets
* Print the Lasso RMSE score on the test sets

<h3>Method 3: Decision Tree </h3>
* Fit Decision Tree to the training set. 
* Print the Decision Tree R squared score on the test sets
* Print the Decision Tree RMSE score on the test sets

<h3>Method 4: Random Forest </h3>
* Fit Random Forest to the training sets. 
* Print the Random Forest R squared score on the test sets
* Print the Random Forest RMSE score on the test sets

## 1. Import dataset "final_data.csv" in pandas 

In [1]:
# Importing pandas package to read the csv file
import pandas as pd
dataset = pd.read_csv("final_data.csv")
print("================")
print("Dataset Features")
print("================")
print(dataset.columns)

Dataset Features
Index(['Unnamed: 0', 'address', 'info', 'z_address', 'bathrooms', 'bedrooms',
       'finishedsqft', 'lastsolddate', 'lastsoldprice', 'latitude',
       'longitude', 'neighborhood', 'totalrooms', 'usecode', 'yearbuilt',
       'zestimate', 'zindexvalue', 'zipcode', 'zpid'],
      dtype='object')


In [2]:
print("====================")
print("Shape of the dataset")
print("====================")
print(dataset.shape)

Shape of the dataset
(11330, 19)


In [3]:
print("========================")
print("Looking into the dataset")
print("========================")
print(dataset.head())

Looking into the dataset
   Unnamed: 0                             address  \
0           2  Address: 1160 Mission Street #2007   
1           5       Address: 260 King Street #475   
2           7     Address: 560 Missouri Street #B   
3           9        Address: 350 Missouri Street   
4          11         Address: 3658 Folsom Street   

                                                info  \
0   San FranciscoSales price: 1300000Sales date: ...   
1   San FranciscoSales price: 750000Sales date: 0...   
2   San FranciscoSales price: 1495000Sales date: ...   
3   San FranciscoSales price: 2700000Sales date: ...   
4   San FranciscoSales price: 1530000Sales date: ...   

                   z_address  bathrooms  bedrooms  finishedsqft lastsolddate  \
0  1160 Mission St UNIT 2007        2.0       2.0        1043.0   02/17/2016   
1       260 King St UNIT 475        1.0       1.0         903.0   02/17/2016   
2        560 Missouri St # B        4.0       3.0        1425.0   02/17/2016   

The data provided is the home prices in the San Francisco. As the info field location is San Francisco in the entire dataset

## 2. Preprocess/Clean the Dataset

In [4]:
# Check the data types of the data set for any incompactible data types for the features
print("================================")
print("Checking the features data types")
print("================================")
print(dataset.dtypes)

Checking the features data types
Unnamed: 0         int64
address           object
info              object
z_address         object
bathrooms        float64
bedrooms         float64
finishedsqft     float64
lastsolddate      object
lastsoldprice    float64
latitude         float64
longitude        float64
neighborhood      object
totalrooms       float64
usecode           object
yearbuilt        float64
zestimate        float64
zindexvalue       object
zipcode          float64
zpid             float64
dtype: object


From the above, the feature "lastsolddate" doesn't have the required data type. Which we need to convert. Either use the astype() or to_datetime() for converting the "lastsolddate" feature

In [5]:
# Look into the data in the zindexvalue feature
print("===========================")
print("Data of zindexvalue feature")
print("===========================")
print(dataset[["zindexvalue"]].head())

Data of zindexvalue feature
  zindexvalue
0     975,700
1     975,700
2   1,277,600
3   1,277,600
4   1,248,000


The "zindexvalue" feature contains the numeric data where the data type for this feature is defined as "object" which we need to convert to "float" data type

In [6]:
# Converting the lastsolddate to datetime data type
dataset["lastsolddate"]=pd.to_datetime(dataset["lastsolddate"])
# Converting the zindexvalue to float
# Getting this error while converting ValueError: could not convert string to float: '975,700' for this 
# replace the ',' with ''
dataset["zindexvalue"]=dataset["zindexvalue"].str.replace(',','').astype("float")
# We can perform convertion in two steps or in one step as above
# dataset["zindexvalue"]=dataset["zindexvalue"].str.replace(',','')
# dataset["zindexvalue"]=dataset["zindexvalue"].astype("float")
# Checking the column types after changing the dtype of the zindexvalue
print("==============================================================")
print("Checking the features data types of the dataset after changing")
print("==============================================================")
print(dataset.dtypes)

Checking the features data types of the dataset after changing
Unnamed: 0                int64
address                  object
info                     object
z_address                object
bathrooms               float64
bedrooms                float64
finishedsqft            float64
lastsolddate     datetime64[ns]
lastsoldprice           float64
latitude                float64
longitude               float64
neighborhood             object
totalrooms              float64
usecode                  object
yearbuilt               float64
zestimate               float64
zindexvalue             float64
zipcode                 float64
zpid                    float64
dtype: object


In [7]:
# Check if there are any missing values in the data set
print("====================================================")
print("Checking if there are any null values in the dataset")
print("====================================================")
print(dataset.isnull().sum())

Checking if there are any null values in the dataset
Unnamed: 0       0
address          0
info             0
z_address        0
bathrooms        0
bedrooms         0
finishedsqft     0
lastsolddate     0
lastsoldprice    0
latitude         0
longitude        0
neighborhood     0
totalrooms       0
usecode          0
yearbuilt        0
zestimate        0
zindexvalue      0
zipcode          0
zpid             0
dtype: int64


The dataset doesn't have any missing values. Now we will extract year from the "lastsolddate" as we dont want the entire date. We'll call extracted new feature as "lastsoldyear" and convert the data type to float to match with the data type of "yearbuilt" fearure

In [8]:
# Extract year from the date column
dataset["lastsoldyear"] = dataset["lastsolddate"].apply(lambda row: row.year)
# Print the "lastsolddate", "lastsoldyear"
print("==================================================")
print("After extraction of the year from the lastsolddate")
print("==================================================")
print(dataset[["lastsolddate","lastsoldyear"]].head())
# Convert the year(which is the lastsolddate) from int to float as the same type as yearbuilt
dataset["lastsoldyear"]=dataset["lastsoldyear"].astype("float")

After extraction of the year from the lastsolddate
  lastsolddate  lastsoldyear
0   2016-02-17          2016
1   2016-02-17          2016
2   2016-02-17          2016
3   2016-02-17          2016
4   2016-02-17          2016


1. Drop the features which are redundant and not useful. The dataset has one unnamed feature as we dont get any information from the feature drop that feature. When queried the features it is identified as "Unnamed: 0". So drop that feature. We may assume that feature contains the ID after removing the null values.
2. The features addrees, z_address, latitude, longtitude, neighborhood, zipcode all talks about the location. As we want to estimate the price based on the neighborhood we will drop remaining features related to location. 
3. The info feature is broken down into features like bedrooms, bathrooms, lastsoldprice, lastsolddate, finishedsqft. It does not create the new feature for lotsize.
4. The feature zestimate is the market price estimated by the zillow.
5. zpid is the zillow property id.

In [9]:
# Droopping the features. We will keep the z_address to find the median with respect to neighborhood
to_drop=["address","info","lastsolddate","latitude","longitude","Unnamed: 0","zestimate","zpid","zipcode"]
dataset=dataset.drop(to_drop,axis=1)

In [10]:
# Check the dataset columns after removing
print("======================================")
print("Features of the dataset after removing")
print("======================================")
print(dataset.columns)

Features of the dataset after removing
Index(['z_address', 'bathrooms', 'bedrooms', 'finishedsqft', 'lastsoldprice',
       'neighborhood', 'totalrooms', 'usecode', 'yearbuilt', 'zindexvalue',
       'lastsoldyear'],
      dtype='object')


Check the unique values in the "usecode" and "neighborhood" features

In [11]:
# Use the value_counts() to check the unique value count for usecode data
print("===================")
print("usecode value_count")
print("===================")
print(dataset["usecode"].value_counts())

usecode value_count
SingleFamily        5803
Condominium         4802
MultiFamily2To4      486
Duplex               146
Townhouse             66
Miscellaneous         17
Apartment              3
Cooperative            3
Mobile                 2
MultiFamily5Plus       2
Name: usecode, dtype: int64


In [12]:
# Use the value_counts() and count() to check the count of unique neighborhood data
print("==============================================")
print("Total count of unique neighborhood value_count")
print("==============================================")
print(dataset["neighborhood"].value_counts().count())

Total count of unique neighborhood value_count
71


Different neighborhood as different price for the house. Find the price for per square feet. To calculate the price per square feet we divide the lastsoldprice with finishedsqft and store the result in the "pps"

In [13]:
dataset["pps"]=dataset["lastsoldprice"]/dataset["finishedsqft"]

In [14]:
# Check the dataset columns after calculating the price per square feet
print("=============================================")
print("Features of the dataset after calculating pps")
print("=============================================")
print(dataset.columns)

Features of the dataset after calculating pps
Index(['z_address', 'bathrooms', 'bedrooms', 'finishedsqft', 'lastsoldprice',
       'neighborhood', 'totalrooms', 'usecode', 'yearbuilt', 'zindexvalue',
       'lastsoldyear', 'pps'],
      dtype='object')


The feature "usecode" has categorical data which we need to encode for this we are using one-hot encoding technique

In [15]:
# Use the get_dummies() to do one-hot encoding
dataset_usecode_enc=pd.get_dummies(dataset["usecode"])
dataset = pd.concat([dataset, dataset_usecode_enc], axis=1)
print("===========================================")
print("Features after one-hot encoding for usecode")
print("===========================================")
print(dataset.columns)

Features after one-hot encoding for usecode
Index(['z_address', 'bathrooms', 'bedrooms', 'finishedsqft', 'lastsoldprice',
       'neighborhood', 'totalrooms', 'usecode', 'yearbuilt', 'zindexvalue',
       'lastsoldyear', 'pps', 'Apartment', 'Condominium', 'Cooperative',
       'Duplex', 'Miscellaneous', 'Mobile', 'MultiFamily2To4',
       'MultiFamily5Plus', 'SingleFamily', 'Townhouse'],
      dtype='object')


The feature "usecode" is no longer needed after one-hot encoding so drop it

In [16]:
# Drop the usecode feature
to_drop=["usecode"]
dataset=dataset.drop(to_drop,axis=1)

Group the neighborhood based on the price per square feet

In [17]:
# mean_value=dataset.groupby(["neighborhood"]).mean()["pps"]
mean_value=dataset.groupby(["neighborhood"]).mean()
neighborhood_pps=mean_value["pps"]

In [18]:
dataset_temp=pd.concat([neighborhood_pps],axis=1)
dataset_temp["neighborhood"]=dataset_temp.index
dataset_temp.columns.values[0]="pps"

Get the median of the neighborhood_pps by using the median()

In [19]:
median_neighborhood_pps=neighborhood_pps.round().median()
print(median_neighborhood_pps)

756.0


In [20]:
dataset1=dataset_temp["pps"] < median_neighborhood_pps
dataset1_value=dataset_temp[dataset1]

In [21]:
dataset2=dataset_temp["pps"] >= median_neighborhood_pps
dataset_2_temp = dataset_temp[dataset2] 

In [22]:
def get_group(x):
    if x in dataset1_value.index:
        return 'lp'
    else:
        return 'hp'
dataset['group'] = dataset["neighborhood"].apply(get_group)

In [23]:
# Check the features of the dataset
print("==========================================")
print("Features of the dataset after adding group")
print("==========================================")
print(dataset.columns)

Features of the dataset after adding group
Index(['z_address', 'bathrooms', 'bedrooms', 'finishedsqft', 'lastsoldprice',
       'neighborhood', 'totalrooms', 'yearbuilt', 'zindexvalue',
       'lastsoldyear', 'pps', 'Apartment', 'Condominium', 'Cooperative',
       'Duplex', 'Miscellaneous', 'Mobile', 'MultiFamily2To4',
       'MultiFamily5Plus', 'SingleFamily', 'Townhouse', 'group'],
      dtype='object')


Use the value_counts() to count the unique values in the group feature

In [24]:
print(dataset["group"].value_counts())

hp    6998
lp    4332
Name: group, dtype: int64


In [25]:
dataset_group_enc=pd.get_dummies(dataset["group"])
dataset= pd.concat([dataset, dataset_group_enc], axis=1)

In [26]:
# Drop the neighborhood and group feature as we one hot encoded into "high_price" and "low_price"
to_drop=["group","neighborhood","pps","z_address"]
dataset=dataset.drop(to_drop,axis=1)

See the final features and the shape of the dataset using the columns and shape attribute

In [27]:
print("=============================")
print("Features of the final dataset")
print("=============================")
print(dataset.columns)
print("==========================")
print("Shape of the final dataset")
print("==========================")
print(dataset.shape)


Features of the final dataset
Index(['bathrooms', 'bedrooms', 'finishedsqft', 'lastsoldprice', 'totalrooms',
       'yearbuilt', 'zindexvalue', 'lastsoldyear', 'Apartment', 'Condominium',
       'Cooperative', 'Duplex', 'Miscellaneous', 'Mobile', 'MultiFamily2To4',
       'MultiFamily5Plus', 'SingleFamily', 'Townhouse', 'hp', 'lp'],
      dtype='object')
Shape of the final dataset
(11330, 20)


In [28]:
X = dataset.drop('lastsoldprice', axis=1)
y = dataset['lastsoldprice']

In [29]:
from sklearn.model_selection import train_test_split 
train_X, test_X, train_y, test_y = train_test_split(X, y)

In [30]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
# Fit knn to the training sets
knn.fit(train_X, train_y)
# Print the score of knn on the train and test sets
print("=======================")
print("Score of the train sets")
print("=======================")
print(knn.score(train_X, train_y))
print("======================")
print("Score of the test sets")
print("======================")
print(knn.score(test_X, test_y))

Score of the train sets
0.24479227962810404
Score of the test sets
0.016590187080833037


## 3. Method 1: Multiple Linear Regression

Implement the backward selection for identifying the potential features

In [31]:
# Import the numpy package to calculate the RMSE
import numpy as np
# Import the mean_squared_error from sklearn.metrics
from sklearn.metrics import mean_squared_error

In [32]:
# Below is the automatic implementation of Backward elimination
def backwardElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j] == maxVar):
                    x = np.delete(x, j, 1)
    regressor_OLS.summary()
    return x

In [33]:
# Bulding the optimal model using Backward Elimination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((11330,1)), values = X, axis = 1)
SL = 0.05
X_Modeled = backwardElimination(X, SL)

In [34]:
# Splitting the dataset into training and test data set
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X_Modeled, y, test_size = 0.3, random_state = 0)
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(train_X, train_y)
print("=================================")
print("Multi Linear Regression R Squared")
print("=================================")
print('Multi Linear Regression R squared": %.4f' % regressor.score(test_X, test_y))

Multi Linear Regression R Squared
Multi Linear Regression R squared": 0.5880


Calculate the RMSE for the Multi linear Regression

In [35]:
# Calculate root-mean-square error (RMSE). Predict the test results
y_pred = regressor.predict(test_X)
lin_mse = mean_squared_error(y_pred, test_y)
lin_rmse = np.sqrt(lin_mse)
print("============================")
print("Multi Linear Regression RMSE")
print("============================")
print('Multi Linear Regression RMSE: %.4f' % lin_rmse)

Multi Linear Regression RMSE
Multi Linear Regression RMSE: 597482.6651


The below train_test_split is used for LASSO, Decision Tree Regression and Random Forest

In [36]:
from sklearn.model_selection import train_test_split 
train_X, test_X, train_y, test_y = train_test_split(X, y,test_size=0.3,random_state=0)

## 4. Method 2: Least Absolute Shrinkage and Selection Operator(Lasso)

In [37]:
from sklearn import linear_model
lasso=linear_model.Lasso(alpha=1000,max_iter=1e4)
lasso.fit(train_X,train_y)
print("==========================")
print("Lasso Regression R Squared")
print("==========================")
print('Lasso Regression R Squared: %.4f'%lasso.score(test_X,test_y))

Lasso Regression R Squared
Lasso Regression R Squared: 0.5691


In [38]:
# Calculate root-mean-square error (RMSE)
y_pred = lasso.predict(test_X)
lin_mse = mean_squared_error(y_pred, test_y)
lin_rmse = np.sqrt(lin_mse)
print("=====================")
print("Lasso Regression RMSE")
print("=====================")
print('Lasso Regression RMSE: %.4f' % lin_rmse)

Lasso Regression RMSE
Lasso Regression RMSE: 611030.0425


## 5. Method 3: Decision Tree

In [39]:
from sklearn.tree import DecisionTreeRegressor
regressor_decision=DecisionTreeRegressor(random_state=0)
regressor_decision.fit(train_X,train_y)
print("==================================")
print("Decision Tree Regression R Squared")
print("==================================")
print('Decision Tree Regression R squared: %.4f'%regressor_decision.score(test_X,test_y))

Decision Tree Regression R Squared
Decision Tree Regression R squared: 0.4190


In [40]:
# Calculate root-mean-square  error(RMSE)
y_pred=regressor_decision.predict(test_X)
lin_mse=mean_squared_error(y_pred,test_y)
lin_rmse=np.sqrt(lin_mse)
print("=============================")
print("Decision Tree Regression RMSE")
print("=============================")
print('Decision Tree Regression RMSE: %.4f' % lin_rmse)

Decision Tree Regression RMSE
Decision Tree Regression RMSE: 709507.0142


## 6. Method 4: Random Forest

In [41]:
from sklearn.ensemble import RandomForestRegressor
forest_reg=RandomForestRegressor(n_estimators=1000,random_state=0)
forest_reg.fit(train_X,train_y)
print("=======================")
print("Random Forest R Squared")
print("=======================")
print('Random Forest R squared: %.4f'%forest_reg.score(test_X,test_y))

Random Forest R Squared
Random Forest R squared: 0.7175


In [42]:
# Calculate root-mean-square  error(RMSE)
y_pred=forest_reg.predict(test_X)
forest_mse=mean_squared_error(y_pred,test_y)
forest_rmse=np.sqrt(forest_mse)
print("==================")
print("Random Forest RMSE")
print("==================")
print('Random Forest RMSE: %.4f'%forest_rmse)

Random Forest RMSE
Random Forest RMSE: 494732.5933


## 7. Test Data Set Prediction Results Comparison Table of Method 1, Method 2, Method 3 and Method 4
* Please fill in your results to the following table

<table>
<thead>
    <tr>
    <th colspan="4">Predict The Bay Area’s Home Prices</th>
    </tr>
    <tr>
    <th> </th>
    <th>Multiple Linear Regression</th>
    <th>Lasso</th>
    <th>Decision Tree</th>
    <th>Random Forest</th>
    </tr>
</thead>
<tbody>
    <tr>
        <th>R Squared Score</th>
        <td> 0.5880 </td>
        <td> 0.5691 </td>
        <td> 0.4190 </td>
        <td> 0.7175 </td>
    </tr>
     <tr>
         <th>RMSE</th>
        <td> 597482.6651 </td>
        <td> 611030.0425 </td>
        <td> 709507.0142 </td>
        <td> 494732.5933 </td>
    </tr>
</tbody>
</table>